arangod/RocksDBEngine/RocksDBDumpManager.cpp

```diff
@@ -0,0 +1,155 @@
+
+using namespace arangodb;
+
+RocksDBDumpManager::RocksDBDumpManager(RocksDBEngine& engine)
+    : _engine(engine) {}
+
+RocksDBDumpManager::~RocksDBDumpManager() { _contexts.clear(); }
+
+std::shared_ptr<RocksDBDumpContext> RocksDBDumpManager::createContext(
+    RocksDBDumpContextOptions opts, std::string const& user,
+    std::string const& database) {
+  TRI_ASSERT(ServerState::instance()->isSingleServer() ||
+             ServerState::instance()->isDBServer());
+
+  // generating the dump context can throw exceptions. if it does, then
+  // no harm is done, and no resources will be leaked.
+  auto context = std::make_shared<RocksDBDumpContext>(
+      _engine, _engine.server().getFeature<DatabaseFeature>(), generateId(),
+      std::move(opts), user, database);
+
+  std::lock_guard mutexLocker{_lock};
+
+  if (_engine.server().isStopping()) {
+    // do not accept any further contexts when we are already shutting down
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_SHUTTING_DOWN);
+  }
+
+  bool inserted = _contexts.try_emplace(context->id(), context).second;
+  if (!inserted) {
+    // cannot insert into map. should never happen
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "unable to insert dump context");
+  }
+
+  TRI_ASSERT(_contexts.at(context->id()) != nullptr);
+
+  return context;
+}
+
+std::shared_ptr<RocksDBDumpContext> RocksDBDumpManager::find(
+    std::string const& id, std::string const& database,
+    std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  // this will throw in case the context cannot be found or belongs to a
+  // different user
+  auto it = lookupContext(id, database, user);
+  TRI_ASSERT(it != _contexts.end());
+
+  return (*it).second;
+}
+
+void RocksDBDumpManager::remove(std::string const& id,
+                                std::string const& database,
+                                std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  // this will throw in case the context cannot be found or belongs to a
+  // different user
+  auto it = lookupContext(id, database, user);
+  TRI_ASSERT(it != _contexts.end());
+
+  // if we remove the context from the map, then the context will be
+  // destroyed if it is not in use by any other thread. if it is in
+  // use by another thread, the thread will have a shared_ptr of the
+  // context, and the context will be destroyed once the shared_ptr
+  // goes out of scope in the other thread
+  _contexts.erase(it);
+}
+
+void RocksDBDumpManager::dropDatabase(TRI_vocbase_t& vocbase) {
+  std::lock_guard mutexLocker{_lock};
+
+  std::erase_if(_contexts, [&](auto const& x) {
+    return x.second->database() == vocbase.name();
+  });
+}
+
+void RocksDBDumpManager::garbageCollect(bool force) {
+  auto const now = TRI_microtime();
+
+  std::lock_guard mutexLocker{_lock};
+
+  std::erase_if(_contexts, [&](auto const& x) {
+    return force || x.second->expires() < now;
+  });
```

Comment:
I find this easier to read, but feel free to keep it the way it is
```suggestion
  std::lock_guard mutexLocker{_lock};

  if (force) {
    _contexts.clear();
    return;
  }
  auto const now = TRI_microtime();
  std::erase_if(_contexts, [&](auto const& x) { return x.second->expires() < now; });
```